File: codes/bpsk_ber.py
#if using termux
import subprocess
import shlex
#end if
import numpy as np
import math
import matplotlib.pyplot as plt
def qfunc(x):
	return 0.5*math.erfc(x/math.sqrt(2))

# SNR values in db
SNR_db = np.linspace(0,9,10)
# SNR values
SNR = 10**(0.1*SNR_db)
#simualtion ber declaration
err_sim = []
#No.of samples
simlen = int(1e5)
#analytical ber declaratiom
err_ana = []

for i in range(0,len(SNR)):
	#AWGN with mean 0 and variance 1
	n1 = np.random.normal(0,1,simlen)
	
	# received  signal
	rx = math.sqrt(SNR[i]) + n1
	#Storing indices of rx<0
	err_ind = np.nonzero(rx < 0)
	#calculating simulation error
	err_sim.append((np.size(err_ind))/simlen)
	#calculating analytical error
	err_ana.append(qfunc(math.sqrt(SNR[i])))
	
plt.semilogy(SNR_db.T,err_ana,label='Analysis')
plt.semilogy(SNR_db.T,err_sim,'o',label='Sim')
plt.xlabel('SNR$\\left(\\frac{E_b}{N_o}\\right)$')
plt.ylabel('$P_e$')
plt.legend()
plt.grid()
#if using termux
plt.savefig('./figs/ee18btech11042/ee18btech11042.pdf')
plt.savefig('./figs/ee18btech11042/ee18btech11042_1.eps')
subprocess.run(shlex.split("termux-open ./figs/ee18btech11042/ee18btech11042.pdf"))
#else
#plt.show()
plt.show()

File: codes/bpsk_ber.m

%%SNR values in db
SNR_db = linspace(0,9,10);
%%SNR values
SNR = 10.^(0.1.*SNR_db);
%%simulated ber declaration
err_sim = [];
%%analytical ber declaration
err_ana = [];
simlen = 100000;
for i = 1:10
    n = normrnd(0,1,[1,simlen]);
    rx = sqrt(SNR(i)) + n;
    
    %%storing indices of rx<0
    idx = rx<0;
    neg = rx(idx);
    %%calculating simulation error
    err_sim = [err_sim,(length(neg))/simlen];
    %%calculating analytical error
    a = 0.5*erfc((sqrt(SNR(i)))/sqrt(2));
    err_ana = [err_ana,a];
end
semilogy(SNR_db,err_ana);

hold on
xlabel('E_b/N_o');
ylabel('p_e');
semilogy(SNR_db,err_sim,'o');
hold off






